FreeRTOS/FreeRTOS-Plus-TCP
# IPv4/single: TCP handles partial segments.

Description
-----------
When using UDP, it is possible to receive data without first copying it. In that case it will return a pointer to a network buffer.

This is the way it works:

~~~c
    struct freertos_sockaddr xAddress;
    socklen_t uxAddrLen = sizeof xAddress;
    uint8_t *pucBuffer = NULL;
    BaseType_t xLength = FreeRTOS_recvfrom( xSocket, &( pucBuffer ), 0, FREERTOS_ZERO_COPY, &( xAddress ), &( uxAddrLen ) );
    if( xLength > 0 )
    {
        /* Use the contents of 'pucBuffer'. */
        /* Release the network buffer that contains 'pucBuffer'. */
        FreeRTOS_ReleaseUDPPayloadBuffer( pucBuffer );
    }
~~~

For TCP sockets, it is possible to peek into the circular reception buffer without copying it. When done, the pointer in the reception buffer must be advanced to mark the data as "read".

Recently, a user Matt Boytim on the [FreeRTOS forum](https://forums.freertos.org/t/releaseudppayloadbuffer-but-for-tcp/13890) asked for a function similar to 'FreeRTOS_ReleaseUDPPayloadBuffer()' for TCP. We wrote this function and named it, not surprisingly 'FreeRTOS_ReleaseTCPPayloadBuffer()'. 

~~~c
    uint8_t *pucBuffer = NULL;
    BaseType_t xLength = FreeRTOS_recv( xSocket, &( pucBuffer ), 0, FREERTOS_ZERO_COPY );
    if( xLength > 0 )
    {
        /* Use the contents of 'pucBuffer'. */
        /* Release the space that was occupied by received data. */
        FreeRTOS_ReleaseTCPPayloadBuffer( xSocket, pucBuffer, xLength );
    }
~~~

This technique of calling FreeRTOS_recv() with the zero-copy flag always seemed to work well. Except, we found out, when using it in an echo server.

~~~c
    BaseType_t xLength = FreeRTOS_recv( xSocket, &( pucBuffer ), 0, FREERTOS_ZERO_COPY );
    if( xLength > 0 )
    {
        FreeRTOS_send( xSocket, pucBuffer, xLength, 0 );
        FreeRTOS_ReleaseTCPPayloadBuffer( xSocket, pucBuffer, xLength );
    }
~~~

When calling FreeRTOS_send(), the driver starts returning the data to the echo client, without sending an ACK for the received data.

The client then sees that the earlier packet must have been dropped and quickly repeats it. The echo-server will send an ACK only after returning the data. These two packets may cross: the repeated data and the ACK.

Still not a big problem, except when the echo-client sends the packet again, but now with more bytes:

![image](https://user-images.githubusercontent.com/20884021/145537846-0621dedb-d1bf-4349-9fae-6b07534290fc.png)

Wireshark marks the repeated packet as "**TCP out of order**", and FreeRTOS+TCP just drops the packet, thinking it is an error.

This is all not fatal, but it leads to time-outs and unnecessary Selective ACK's.

This PR repairs that: when receiving a TCP segment that was partially seen already, it will add the unseen bytes and acknowledge it properly. After this PR, it is OK to use zero-copy perception in an echo-server.


Test Steps
-----------
Create an echo server using zero-copy reception, as described above.

Related Issue
-----------
The communication will not run smoothly.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
